### ISSUE TITLE
Run test methods within a fixture in parallel

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#164]
Currently, we only have implemented parallel execution at the fixture level. It should be possible to run individual methods of a fixture in parallel as well as individual test cases for a parameterized method.

We should be able to specify the exact level of parallelism on the fixture or method.

This issue was previously part of #66

**Comment by agray** [comment]
I would have thought this was relatively easy to implement given the capability exists in the old MbUnit.

Parallel execution of Test cases is an excellent feature of MbUnit and JUnit. It is a major gaping hole in my opinion that it is not also part of NUnit.

**Comment by CharliePoole** [comment]
Well it's in the spec and scheduled for post-3.0 implementation. If it were so easy, we probably would have done it. Not sure what connection you see with mbunit. The fact that they had it doesn't help us.

**Comment by agray** [comment]
My understanding of the connection with MbUnit is two fold.

1. @rprouse was a key contributor to MbUnit in a past life.
2. MbUnit version 3 has parallel test case execution just like a JUnit 4 does... and NUnit 3 should have.

If Compaq could find a way to build a team in the 1980's to reverse engineer an IBM machine, surely there is a way to bring this functionality into NUnit.

**Comment by CharliePoole** [comment]
This is getting a bit tiresome. A few points, already stated but apparently missed...

1. There is a plan to implement what you are asking for.

2. We decided as a team to schedule it at a certain point.

3. When it is scheduled has primarily to do with our assessment of the value of the feature as compared with other features.

4. There are several people on the NUnit team who could implement the feature.

5. They would be able to implement it, depending on priorities, out of their heads and would not need to copy it from anywhere.

I find your talk of "reverse engineering" very disturbing. Open source is only possible in a context where copyright is respected. So if you are suggesting that we might ignore the mbunit licensing terms, you are way off base.

**Comment by rprouse** [comment]
While I did contribute many patches to MbUnit and used it for years, I was
never a key contributor. I wouldn't want my status to be misrepresented :)

As for reverse engineering, there isn't really any use here. NUnit works
entirely different than MbUnit did. We will get to this issue in due time,
but are approaching it cautiously because other similar issues might change
the way we need to implement this or even conflict.

**Comment by CharliePoole** [comment]
A much more tactful comment than mine!

Our current priorities in the world of parallel execution are:
1. Parallel process execution
2. Exclusion groups
3. Parallel method (in one fixture) execution.

This seems to represent the order of greatest usefulness to users.

I'll also mention that marking something as post-3.0 does not necessarily mean the feature comes at a later time than it would if we made it part of 3.0. Rather, it may mean that 3.0 comes at an earlier point in time.

**Comment by rprouse** [comment]
When we do this, we are going to have to decide how to handle setup and teardown commands. The easiest option would likely be to construct the test class for each test so that there is no contention for data.

**Comment by CharliePoole** [comment]
I would favor making that an option at some point anyway, but I'm not sure whether it's a run option or a fixture-by-fixture (attributed) option or both.

**Comment by julianhaslinger** [comment]
Hi! 
The addition of this feature to the next version of NUnit would be great, since it is the only "blocker bug" that prevents me from switching to NUnit. Is this feature still planned for 3.4?

**Comment by CharliePoole** [comment]
@julianhaslinger NUnit 3.4 will be out at the end of the month, so no, this feature will not be included.

FYI, this issue is in our Backlog milestone (or pseudo-milestone) rather than 3.4 because we are following a practice of only adding a small number of key defining issues to each numbered milestone in advance. 

The next milestone, 3.6, is scheduled to drop in 3 more months, which probably sounds discouraging to you. {-( However, if you see this issue being merged to master, you will be able to get an earlier drop from our MyGet feed.

**Comment by rprouse** [comment]
@julianhaslinger this won't be in 3.4 which is probably out in just over a week, but it is one I would like to see done soon, so your vote helps :+1: 

As an aside, which test framework are you using that supports parallel method execution? I thought [XUnit only allowed running tests in parallel down to the Class level](https://xunit.github.io/docs/running-tests-in-parallel.html) too. Am I mistaken? I don't use XUnit much :smile:

**Comment by julianhaslinger** [comment]
@CharliePoole Thanks for the answer! Ok, this sounds a bit discouraging, though. However, I will wait for the next releases (or earlier drops). Please (still) consider this feature in the next version. 

@rprouse I'm using MbUnit/Gallio right now. The project pretty much died and now I'm going to switch to NUnit.

**Comment by CharliePoole** [comment]
@julianhaslinger Is it a question of running time for you? Do you have some numbers on the distribution of test cases within fixtures? I ask because my assumption has been that the feature gives little to users beyond what they get from parallel fixtures. If that's wrong, it could change its relative priority.

**Comment by julianhaslinger** [comment]
@CharliePoole Exactly, our actual problem is the running time of the test cases. We would like to use NUnit for our automated / regression tests (including Selenium / WebDriver) and since some of our test classes have 200+ test cases, it's currently a real pain to execute the whole test suite with NUnit. 

Comparisons (within a smaller test set of our regression tests) have shown that a possible NUnit solution will be executed in about 1,5 hours whereas the current MBunit solution will be executed in 0,5 hours (same test set). 

However, @CharliePoole, I'm aware of the fact that we might be able to reduce execution time by splitting bigger test classes into smaller ones, but this would somehow undermine the initial idea of our test class hierarchy.

**Comment by ChrisMaddock** [comment]
@julianhaslinger - Just to confirm, are you sure you're running time is CPU bound, and not memory bound?

We experienced the same issue when we first tried switching to NUnit 3 - run time became significantly longer. This was due to NUnit maxing out the available memory however - there's an issue with the current version which holds on to all test objects until the test run is complete. We're currently running a patched version internally - I wanted to get a PR in for 3.4, but not sure I'll have time at this point. :-(

Apologies if this isn't it, but thought it was worth a mention, as Selenium can use it's fair share of memory. :-)

**Comment by julianhaslinger** [comment]
@ChrisMaddock Thanks for the input! :+1: We will have a look at it soon and I will come back to you then.

**Comment by rprouse** [comment]
@ChrisMaddock I would love to see your memory fix in 3.4. If you want to put up a PR, we can help you clean it up if it isn't ready for production yet :+1:

**Comment by ChrisMaddock** [comment]
@rprouse - This initial commit https://github.com/nunit/nunit/pull/1367/commits/5f98ae51025f7af8244abd4367d1f47260874dfc in PR #1367 frees things up properly for us, in a patched v3.0.1.

You'll see in the PR that it was moved to `OneTimeTearDown` before merge - I don't yet know if that move caused it not to work in v3.2.1, or if there was another change working which reintroduced the retention - I may well not have retested after moving.

Will try and re-test it and throw it up tomorrow, would be great for us to go back to core releases as well.

**Comment by julianhaslinger** [comment]
@ChrisMaddock I had a look at the memory consumption during a regression test run and found out that the execution of test runs is not memory-bound. I guess we really have to wait for a version of NUnit that can handle the parallel execution of individual methods inside each class.

**Comment by bicarbon8** [comment]
just to add to what @julianhaslinger said, I've seen exactly the same issue in using NUnit with Selenium tests. I had to write a custom wrapper that runs individual instances of nunit-console.exe against a single test as scraped from a passed in assembly so I could run parallel tests for now. This is not ideal though because it results in many .xml output files being generated (one for each test run) rather than a single .xml output and it also means I cannot make use of _OneTimeSetUp_ and teardown methods and instead must rely on external semaphores and environment variables to control execution flow where I only want to run things once. I've been following the promise of having parallel execution in NUnit for a long time now (years) and was very disappointed at how the feature was rolled out in 3.0 (only partially supported and requiring some real digging in the issues and release notes to discover not everything was actually supported). I even went so far as to investigate what would be involved in moving this forward myself, but unfortunately it appears that NUnit's design is working against the implementation of parallel execution at the test method level, I've no idea how you'd actually like to handle the potential for non-threadsafe code (i.e. should it be left to the person using NUnit in parallel to ensure all test methods properly reference any externals in a threadsafe manner or should NUnit itself handle execution of each method in it's own domain while somehow trying to still only run the onetime setup and teardown type methods only once) and since I already have a workaround in place I could not justify the time in implementing this feature myself (which is probably for the best because I've been burned before when working with teams on Git to help them implement parallel execution). Anyway... sorry for rambling... I understand that the priority for Unit Testing of C# code probably doesn't rely on method-level parallelism as much, but please do know that for those of us using NUnit for Selenium-based tests it would be a HUGE improvement. Thanks in advance! :smile:

**Comment by CharliePoole** [comment]
Thanks for your comments. I think I'll ramble on a bit myself here...

It helps to understand what people need and why they need it. Not being a web developer myself, I tried to listen to Selenium users and focus on what they wanted. What that seemed to be was parallel fixtures with ordering among the test cases. I fully grasp that different people will want different things, but when both groups present themselves as expressing "what Selenium users need" it's a bit confusing. Can you outline for me what sorts of web tests might want method-level parallelism as opposed to those folks who want to run methods in a pre-defined order? I think I could guess, but I'd rather have it from a user.

Regarding your workaround... did you consider using separate test assemblies? NUnit would be glad to run each one in parallel in a separate process. Of course that doesn't eliminate the need to synchronize access to things like files.

Regarding thread-safety - we never planned for our parallel implementation to take responsibility for thread-safety. The only thing I was expecting to guarantee is that NUnit won't __create__ any thread-safety problems if your methods are already thread-safe. That in itself turns out to be non-trivial.

**Comment by bicarbon8** [comment]
Thanks for the response. I completely understand your comments about
ordering of test execution in a fixture and splitting tests into separate
assemblies.

I am already a NUnit user and have been for years (and JUnit before that)
so the idea of relying on ordering of tests in a fixture wasn't even a
consideration in my test design. Each test tries to be a fully encapsulated
action or set of actions within the site with no expectation on order of
execution. The design of the site I'm testing also means that many tests
can take up to 20 minutes to execute (much of this time is spent in waiting
for thing to happen on the site) and the type of testing is covering
multiple similar scenarios so base test classes are utilised to reduce code
duplication and increase maintainability by way of using common base class
methods. The tests are organised into separate projects based on the dev
team ownership of that area (so we do use separate assemblies, but there
are many tests per assembly and my test wrapper will handle execution of
all tests in all passed in assemblies in parallel as it was created before
the first official release of NUnit 3 and I wanted to ensure all tests had
the potential for being executed simultaneously given enough threads
available).

Essentially, you can imagine that one assembly might have 10 tests and
another 100 tests, but the assembly with 10 has tests taking 10 minutes
each whereas the assembly with 100 has tests taking 1 minute each. If I
have a machine capable of running 110 parallel threads (which is actually
part of our infrastructure design) then I would expect the tests to finish
in 10 minutes, not in 100 minutes.

Hopefully this helps explain... sorry if things still aren't totally clear,
but the overall point is that method level parallelism is something I've
seen missing in other test libraries (prspec in ruby for example) and when
doing analysis of the performance improvements to be had by adding it i
found it can be a major positive change. See the following as an example:
https://github.com/bicarbon8/rspec-parallel/wiki/Examples
On 1 Jul 2016 00:45, "CharliePoole" <notifications@github.com> wrote:

Thanks for your comments. I think I'll ramble on a bit myself here...

It helps to understand what people need and why they need it. Not being a
web developer myself, I tried to listen to Selenium users and focus on what
they wanted. What that seemed to be was parallel fixtures with ordering
among the test cases. I fully grasp that different people will want
different things, but when both groups present themselves as expressing
"what Selenium users need" it's a bit confusing. Can you outline for me
what sorts of web tests might want method-level parallelism as opposed to
those folks who want to run methods in a pre-defined order? I think I could
guess, but I'd rather have it from a user.

Regarding your workaround... did you consider using separate test
assemblies? NUnit would be glad to run each one in parallel in a separate
process. Of course that doesn't eliminate the need to synchronize access to
things like files.

Regarding thread-safety - we never planned for our parallel implementation
to take responsibility for thread-safety. The only thing I was expecting to
guarantee is that NUnit won't *create* any thread-safety problems if your
methods are already thread-safe. That in itself turns out to be non-trivial.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/nunit/nunit/issues/164#issuecomment-229819324>, or mute
the thread
<https://github.com/notifications/unsubscribe/ACNsyoczCUV9L7kbF2SHB7lLsyBKlrv9ks5qRFUJgaJpZM4CUZ8r>
.

**Comment by CharliePoole** [comment]
Thanks for the info... it's helpful to see a variety of perspectives.

**Comment by MaceWindu** [comment]
My 5 cents.

In our project we have ~25 test assemblies and parallel test execution of tests per assembly and per fixture already greatly reduce execution time. Parallel execution of tests within fixture will improve it even more. Now we even split most time-consuming fixtures into several fixtures to speed up execution, but it will be much better if nunit will run fixture tests in parallel.

What I really want to see in nunit is a test parallelization on test case level, beause splitting such tests into fixtures is not always convenient and lead to code duplication. We have several tests that could take up to several thousands test-cases (each case execution is fast but in total it takes minutes) or just dozens of cases (but each case is slow) and for now such tests are main show-stoppers for us.

**Comment by CharliePoole** [comment]
Thanks for the input... it's an area we are going to address in the next release along with the competing but complementary issue of making sure that some tests don't run in parallel with one another!

My earlier question was more about how people are writing tests, particularly when driving a web application. Some folks insist they want each method to be a sequenced step and the fixture to represent a series of such steps. This is of course non-parallel by definition. Other people appear to write very similar tests without sequencing the tests. My guess is that the latter group write bigger test methods, with a series of actions and asserts in each one.

**Comment by julianhaslinger** [comment]
@CharliePoole Great news! Thank you so much!

**Comment by pablomxnl** [comment]
In my opinion test cases should be isolated and to not depend on other test cases or the execution order. If people still want to run some tests single threaded they could still use [Parallelizable(ParallelScope.None)] and/or OrderAttribute.

This issue being implemented will allow to run test suites faster versus a local webdriver grid or a Saas provider such as BrowserStack or SauceLabs. Currently I try to limit amount of test methods on a test class to the number of available webdriver nodes or browser sessions, so the grid is fully used. Otherwise if a test has 8 test methods and my grid has 8 browser sessions available only one will be used because all test methods of the class will run one at a time.

**Comment by CharliePoole** [comment]
@pablomxnl Absolutely agree. As a coach or project lead, I have always pushed that idea very hard. However, I'm not wearing either of those hats here. I'm a software provider with users who ask for things. If they ask for truly egregious stuff, I just say no. But if they ask for things that are reasonable in some context, but not others - even not most - I give it consideration.

In the case of web sites, we are almost always __not__ talking about unit tests. Higher level functional tests often require sequencing. Whether it is done via a set of steps in the test or a series of methods is an implementation detail for us and a matter of convenience for the user. Personally, I  think we probably need something called a [Step] that sits inside a [StepFixture] class, or some similar names, so we can get all this ordering/sequencing stuff moved away from unit testing. Maybe I'll have time to do that before the web dies. :-)

Anyway, it's always interesting to learn how people actually use your software.

**Comment by julianhaslinger** [comment]
@CharliePoole, please don't get me wrong by asking this (just want to be sure): Will this feature be added to the next version (3.5, due by September 29, 2016)?

**Comment by CharliePoole** [comment]
While I could pretend to predict the future, that would only satisfy you until September. :-)

Let me answer by explaining how we "schedule" features.

We are currently trying (since start of year) an approach that has releases coming out once per quarter. As we get better at releasing, we may increase the pace. Eventually, we may be able to do continuous deployment.

In a volunteer Open Source project, there is no fixed amount of effort availble per month. We can use the "yesterday's weather" approach, but it turns out that the amount of time people have to spend on NUnit, as well as the number of people volunteering, is quite variable.

As a compromise, we select a small number of issues that are key to a release and add them to the milestone in advance. My preference is to add no more than about 25% of what we may hope to get done. The majority of issues in a release are only added to that milestone after they are done or at best when somebody commits to work on them. You generally will not find open issues in our 3.5 milestone without somebody assigned to them, although I do it occasionally if it looks like something is blocking other work.

So, the positive side of what we do is that we can virtually guarantee that the release will come out on time. The negative side is that we can't tell you what will be in it. :-(

For this particular issue... I've given it high priority in order to tell our developers that it's important. Somebody has to be free to work on it and have enough background for a project of this scope and difficulty. If somebody with the right background picks this up in the next few weeks, I would guess that it can be done by the release. As a guideline, I think it would take me about a month, while working on some other things as well, to work my way through this.

Sorry there's no more definitive answer but such an answer would necessarily be a lie!

**Comment by tomersss** [comment]
@CharliePoole  is there any update considering this feature?

**Comment by KPKA** [comment]
@CharliePoole is there any update to this issue?

**Comment by rprouse** [comment]
@KPKA we haven't started work on this yet, so there is no update. For users with ZenHub installed, you will see this issue move from Backlog to ToDo then In Progress as we start working on it.

**Comment by CharliePoole** [comment]
For non-users of Zenhub, you will be able to tell when somebody picks it up by watching who is assigned to the issue.

**Comment by GitSIPA** [comment]
@CharliePoole @rprouse 

Hi!

Is there any news according to this issue?
Whether it will be in the next release or if it is planned in one of the upcoming milestones?

At the moment this is the only issue that is preventing us from switching to NUnit.
Would be kind if someone could confess himself to get to work on this in the near future since there are many people affected I guess.

Hope to hear from you soon

**Comment by rprouse** [comment]
@GitSIPA what test framework are you using that allows you to run test methods in parallel?

To answer you question, nobody has started working on this issue, so there is no ETA, but we will take your vote into account as we are deciding on what to work on next.

**Comment by agray** [comment]
+1,  Agree with @GitSIPA.  This is really important functionality

**Comment by chesser** [comment]
@GitSIPA +1
@rprouse We are using mbunit, but this framework isn't supporting now. And we have some issues with mbunit.

**Comment by GitSIPA** [comment]
@rprouse We are also using MbUnit at the moment, but since the support is not given for the new Visual Studio versions, we considered switching to NUnit.

**Comment by CharliePoole** [comment]
On issue #1921 @jnm2 asked "What stands between us and having parameterized test cases run in parallel? Can I contribute?"

Second question first: For sure! We'd love your help.

And to the first question:

1. We execute "WorkItems", which currently map one-to-one to tests. I think we need to treat OneTimeSetUp and OneTimeTearDown for a fixture as separate WorkItems as a preliminary step. Issue #1096 deals with this.

2. We need to have some way to make a WorkItem depend on other WorkItems, so that it is only scheduled to execute when all those depended on items are complete. We currently do that in a very simplistic way using a CountDown but we need something more general. The first use of this, of course, would be to have OneTimeTearDown depend on completion of all the tests. It would have other uses later, when we implement dependency among tests. I picture this as a queue of pending workitems, but there may be other ways to implement it.

3. With those two things out of the way, we could start working on the main issue itself: scheduling tests to run rather than running them on the same thread that did the OneTimeSetUp for the fixture.

If you'd like to work on this, you'll have to get very familiar with some of the internals of how NUnit actually dispatches tests for execution. I'll be glad to help you do that.

**Comment by jnm2** [comment]
What planning has been done so far?
My needs are to run enough TestCaseSource cases in parallel to saturate the CPU. I have no setup, teardown or ordering requirements. Other people will need those to be taken into account though.

I would expect TestCaseSource execution to be parallelized, so two methods in the same fixture that used different TestCaseSources would execute the enumerations in parallel. Also seems it would be nice if two methods used the same TestCaseSource if it was only executed once- thoughts on this?

**Comment by CharliePoole** [comment]
I'll repeat a story I've been having to tell a lot lately. :smile:

NUnit first loads tests (aka discovers, explores) and then executes them one or more times, depending on the type of runner.

Your TestCaseSource is used in loading tests, not in running them, and we don't even consider it as part of your test - although you may have it in the same class as your test.

IOW, your test cases don't "use" your testcase source, rather NUnit uses the source to create the test cases. Of course, they can't do anything until created. Make sense?

As it happens, the creation of tests from the source is done sequentially. If you have some code being executed in the source that makes that a problem, my first guess would be that you are doing too much in the test case source. For example, you don't want to create instances of classes in your test case source, rather you want to store parameters that will be used to create them. Unfortunately, this could be a whole book chapter. :smile:

So, what this issue is about is __running__ test cases in parallel. That means all test cases under a fixture, unless you marked some as non-parallelizable. Simple, non-parameterized tests as well as individual cases of a parameterized test would be treated equally in this.

It would be an interesting experiment to allow parallelization of tests in fixtures where there is no OneTimeTearDown (I think OneTimeSetUp is OK). It might work easily if you could correctly make that determination. The determination is harder than you would think, however, because OneTimeTearDown can include both inherited methods and global ActionAttributes.

**Comment by julianhaslinger** [comment]
@jnm2 Did you already have the chance to tackle this specific issue?

**Comment by jnm2** [comment]
@julianhaslinger No. It's on my list after my two more immediate issues, 1885 and 1933. If you want to tackle it, all the better!

**Comment by CharliePoole** [comment]
@julianhaslinger Since we don't know one another, please excuse me for saying that this is a tough one to take on. Make sure you dig into how things are being done now and why - the latter is not always obvious I'm afraid. See my comment at https://github.com/nunit/nunit/issues/164#issuecomment-265267804 for a breakdown into separate PRs I'd like to see in order to fix this. If you come up with a simpler approach, post about it before coding it, so we can weigh future considerations against the obvious YAGNI involved in predicting the future.

**Comment by julianhaslinger** [comment]
@CharliePoole @jnm2 Hi guys! 

It's not that I wanted to start implementing this missing feature, but merely to know about its progress. As I can tell from your answers (above), there has been no progress (?) regarding this particular feature request.

**Comment by chris-smith-zocdoc** [comment]
@CharliePoole One of the potential issues I came across while researching the code was how Fixtures are handled right now. Currently all WorkItems have access to the [same Fixture instance](https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Internal/Commands/OneTimeSetUpCommand.cs#L70), which means that if the WorkItems were to be executed in parallel, they would be accessing a single instance of the test class. This is problematic because of the race conditions it creates when tests rely on Setup or Teardown to set instance level fields. 

One solution to this would be to have each WorkItem operate on a new instance of the test class, but this will likely complicate how Fixture setups behave since there would no longer a single instance of the fixture

**Comment by jnm2** [comment]
> IF we were creating a new fixture per instance, we would simply run "one" time setup more than once. Of course, that would hit hard any users who are making use of statics.

I think it's reasonable to require that there is no static state if you opt to run in parallel, to the extent that it's your fault if something breaks for mixing static state and parallelism. Luckily all this code will be under test so it should be hard to miss :D

> Nevertheless, I guess it's time to do something. I'm going to take this on for the first quarter, at least to the extent of running some experiments and possibly coming up with a sensible breakdown of tasks that others can follow up on.

Count me in for that.

**Comment by agray** [comment]
Me too!

**Comment by CharliePoole** [comment]
Running some timing tests, I discovered something interesting. There is an ambiguity in calling test cases parallelizable. This may also apply to fixtures and other suites, but it's less obvious. 

Here's the thing: if a method is completely non-parallelizable, that means it cannot run in parallel with __any other method__. OTOH, if it's only non-parallelizable within it's fixture then it can't run in parallel with __any other method in the same fixture.__

I propose we handle this distinction as follows:

1. If a test method has no Parallelizable attribute, it should run on the same thread as its containing fixture, provided no other attribute (e.g. Apartment) calls for it to be run on a different thread

2. If it has a Parallelizable attribute with a scope of Self, or if a higher level suite specifies a scope of Children, then the test method is run in parallel with the other methods in the fixture.

Thoughts @nunit/core-team ?

**Comment by CharliePoole** [comment]
Left out above:

3. If it has a Parallelizable attribute with a scope of None, then it runs on the non-parallel queue.

**Comment by ChrisMaddock** [comment]
@CharliePoole - your three points make sense, although I'm not clear on what the alternative would be that you're suggesting we don't do. Could you clarify?

**Comment by CharliePoole** [comment]
We currently treat not having an attribute the same as having one with ParallelScope.None. That gets them queued on the non-parallel queue if I enable dispatching them and the test run slows significantly.

**Comment by CharliePoole** [comment]
PR #2011 takes some initial steps toward implementing this feature but contains a #define that forces test cases to run on the thread of their containing fixture. This comment documents what I think needs to be done in order to remove that limitation.

Currently, a fixture's OneTimeTearDown runs on the thread that was used to finish the last test case executed. If test cases run in parallel, we can't predict which test case that will be. It might be a thread with different characteristics from those required for the fixture. For example, if the last test to finish is running in the STA, that's what will be used to run the OneTimeTearDown, even if the OneTimeSetUp ran in an MTA. In many cases, that might not cause any problem but in some it could.

This might actually be a problem with higher-level fixtures as well, but the introduction of parallel test cases means that there will be many more opportunities for a mismatch that causes an error.

So, in order to ensure that fixture teardown takes place on the right sort of thread, we can no longer run it on the thread of the last test to execute. Instead, the dispatcher will need to maintain information about all pending teardowns and be notified to run those teardowns on the proper thread at the proper time. Figuring out how to do that is the "design" phase of this feature.

**Comment by bicarbon8** [comment]
for anyone who's been following this thread since the beginning (back in 2014) and doesn't want to or is unable to implement their own workaround while awaiting this feature addition, I've just stumbled across an implementation using NUnit 2.6.3 available on CodePlex that seems pretty straightforward to use (I've verified it works in running our functional tests in multiple parallel threads. 

http://cpntr.codeplex.com/

apologies in advance @CharliePoole if this message is a bit orthogonal to the recent discussions on this thread, but if anyone else has been waiting the past 3 years for this feature addition based on the promises set forward for NUnit 3 (way back in the early days), I think it might offer a solution until you guys manage to work out the design issues (sounds like you're closing in on the solution).

**Comment by chris-smith-zocdoc** [comment]
@CharliePoole 

> So, in order to ensure that fixture teardown takes place on the right sort of thread, we can no longer run it on the thread of the last test to execute. Instead, the dispatcher will need to maintain information about all pending teardowns and be notified to run those teardowns on the proper thread at the proper time. Figuring out how to do that is the "design" phase of this feature.


Do we need the dispatcher to know of the pending teardown items upfront, or can we dispatch them as they become available? More concretely, where CompositeWorkItem currently calls  [PerformOneTimeTearDown](https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Internal/Execution/CompositeWorkItem.cs#L342-L351) can we use the dispatcher to dispatch a new unit of work onto the correct work shift?

**Comment by CharliePoole** [comment]
@chris-smith-zocdoc Yes, that's exactly what I'm doing. I created a new type of work item, `OneTimeTearDownWorkItem`, which is nested in `CompositeWorkItem` and is dispatched when the last child test is run. Later on, we might look at some efficiencies when the OneTimeSetUp and all the tests have been run on the same thread.

**Comment by gzak** [comment]
Haven't read everything too carefully, but one thing I'd like to request is that the parallelism be "smart", especially for I/O bound tests. For example, if you have 10 threads executing 100 parallelizable tests, it shouldn't be the case that the 10 threads sit and wait for the first 10 to complete before moving on to the next 10. If the first 10 tests start awaiting very long-running I/O tasks, then the threads should be free to move on to other tests. When the I/O tasks complete, threads will continue them as they free up.

Basically, I'm asking for smart throughput management for I/O bound tests that make extensive use of async/await. This is our number one bottleneck in tests, by far.

**Comment by CharliePoole** [comment]
@chris-smith-zocdoc In fact, that's pretty much what I'm doing. I'm essentially using the existing countdown mechanism to trigger dispatch of a one time teardown task. The trick is to get it dispatched on the proper queue.

@gzak Bear in mind that the mechanism for parallel execution already exists. It depends on workers independently pulling tasks rather than having a controller that pushes tasks to workers. So if one worker is busy with a task for a while, other workers will continue to execute other tasks independently. The trick is to set the number of workers based on the nature of the tests being run. NUnit does fairly well by default with normal, compute-bound unit tests but other sorts of tests may require the user setting an appropriate level of parallelism.

**Comment by LirazShay** [comment]
Can someone explain me how does it work?
I have a test class
When I run this tests NOT in parallel - all tests passed
But when I run them with   [Parallelizable(ParallelScope.Children)]
So they do run in parallel (multiple methods in the same class)
but for some reason some tests are failing.
I have some instance fields in this class and it seems that those fields are shared between threads
Am I right?
Do you create only 1 instance of that class and call the methods concurrently ?
@CharliePoole

**Comment by CharliePoole** [comment]
You figured it out! Yes, all tests in a fixture use the same instance. This is historical with NUnit, which has always worked that way. You must choose between running the test cases in parallel and having any state that is modified per test. There is no way around it currently.

That said, if you have a decent proportion of fixtures, simply running fixtures in parallel can give you good performance.

**Comment by agray** [comment]
Is is possible to have the [Parallelizable(ParallelScope.Children)] in the
AssemblyInfo.cs file?

Has anyone seen that working?

**Comment by jnm2** [comment]
@array Yes, in fact that's the only reason I have an AssemblyInfo.cs now.

**Comment by agray** [comment]
Hi Joseph,

What parallelism line are you adding to your AssemblyInfo.cs file?

I'd love to know what works.

Cheers,

Andrew

**Comment by jnm2** [comment]
@agray The one you asked about:

```c#
[Parallelizable(ParallelScope.Children)]
```

**Comment by MaceWindu** [comment]
don't forget target
```cs
[assembly: Parallelizable(ParallelScope.Children)]
```

**Comment by tparikka** [comment]
@LirazShay I use NUnit to drive Selenium tests, and was using fixture-level fields to hold things like references to user accounts and to the Selenium WebDriver instance that I was working with and so was unable to run tests in parallel in-fixture. The way I worked around that was to write a "factory" (I use quotes because I'm not sure it's the right term) that implements IDisposable that for each test encapsulates all my test needs and cleanly tears them down at the end of the test with no need for [TearDown] or [OneTimeTearDown] kind of like so:

     public class TestFactory : IDisposable
        {
        // Instantiate a new SafeHandle instance.
        private readonly System.Runtime.InteropServices.SafeHandle handle = new Microsoft.Win32.SafeHandles.SafeFileHandle(IntPtr.Zero, true);

        // Flag: Has Disposed been called?
        private bool disposed = false;

        public TestFactory()
        {
            this.UserRepository = new List<UserAccount>();
            this.DU = new DataUtility();
        }

		// A list of users created for this test
        public List<UserAccount> UserRepository { get; private set; }

		// A very simple data layer utility that uses Dapper to interact with the database in my application 
        public DataUtility DU { get; private set; }

		// Gets a new user and adds it to the repository
        public UserAccount GetNewUser()
        {
            var ua = new UserAccount();
            this.UserRepository.Add(ua);
            return ua;
        }


        public void Dispose()
        {
            this.Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (this.disposed)
            {
                return;
            }

            if (disposing)
            {
				// Deletes all user accounts created during the test
                foreach (UserAccount ua in this.UserRepository)
                {
                    try
                    {
                        ua.Delete();
                    }
                    catch (Exception)
                    {
                        // Take no action if delete fails.
                    }
                }

                this.DU.DeleteNullLoginFailures(); // Cleans up the database after tests
                Thread.Sleep(1500);
            }

            this.disposed = true;
        }
    }

Then, within a test I can do this:

	[TestFixture]
    public class UserConfigureTests
    {
        [Test]
        public void MyExampleTest()
        {
            using (TestFactory tf = new TestFactory())
            {
                var testUser = tf.GetNewUser();
				
		tf.DU.DoSomethingInTheDatabase(myParameter);
				
		// Test actions go here, and when we exit this using block the TestFactory cleans
		// up after itself using the Dispose method which calls whatever cleanup logic you've written into it
            }
        }
	}

This way, I can avoid a lot of code duplication, and if I ever need to change the dependencies of my test I just do it once in the factory. If anyone has feedback on the strategy I took I'd appreciate it!

**Comment by jnm2** [comment]
@tparikka I highly recommend exactly that approach myself.

**Comment by agray** [comment]
I have the following in my AssemblyInfo file:

[assembly: Parallelizable(ParallelScope.Children)]
[assembly: LevelOfParallelism(16)]


Do I need the LevelOfParallelism attribute as well anymore?

**Comment by jnm2** [comment]
I haven't looked into using `LevelOfParallelism` yet. [It defaults to the number of cores you have.](https://github.com/nunit/docs/wiki/LevelOfParallelism-Attribute)

If your tests are not CPU-bound, higher makes sense. But as always with perf- the answer is so dependent on your scenario that it's better to measure rather than guess.

**Comment by masaeedu** [comment]
@CharliePoole I'm using `TestcaseSource`, but it looks like the resulting testcases aren't actually being executed in parallel. Is something like this expected to work:

```
    [TestFixture]
    class Deserialization
    {
        public static IEnumerable<TimeSpan> ShouldDeserializeAllCases() => Enumerable.Repeat(0, 5).Select(x => TimeSpan.FromSeconds(2));

        [TestCaseSource("ShouldDeserializeAllCases"), Parallelizable(ParallelScope.Children)]
        public void ShouldDeserializeAll(TimeSpan t)
        {
            Thread.Sleep(t);
            Assert.AreEqual(1, 1);
        }
    }
```

The overall time taken is 10 seconds instead of ~2.

**Comment by ParanoikCZE** [comment]
I'll think that there are no children, so in this case, you could better use
[Parallelizable(ParallelScope.All)]
or move your attribute on class level.

**Comment by masaeedu** [comment]
@ParanoikCZE Thanks. I'm actually flying blind with respect to what that attribute means, so I've tried all enum values on there. Regardless of which of `All`, `Children`, `Fixture` or `Self` I use, I get a 10 second execution time (at least in Visual Studio).

**Comment by ParanoikCZE** [comment]
Try this is source of inspiration :)

    class Deserialization
    {
        public static IEnumerable<TestCaseData> ShouldDeserializeAllCases
        {
            get
            {
                for (int i = 1; i <= 5; i++)
                    yield return new TestCaseData(TimeSpan.FromSeconds(i)).SetName($"Thread_worker_{i}");
            }
        }

        [TestCaseSource(nameof(ShouldDeserializeAllCases)), Parallelizable(ParallelScope.Children)]
        public void ShouldDeserializeAll(TimeSpan t) => System.Threading.Thread.Sleep(t);
    }

**Comment by masaeedu** [comment]
@ParanoikCZE Thanks again. I tested this out in Visual Studio and the visualization is much clearer, but the tests are still running sequentially. Easier to see this if you use a constant sleep for each testcase instead of increasing steps.

**Comment by ParanoikCZE** [comment]
Try adding [assembly: LevelOfParallelism(5)] into AssemblyInfo, I think there is some default value, but maybe it doesn't work for you somehow. Anyway, I'm out of ideas. :)
